src/rulesets/Base/array.jl

```diff
@@ -39,6 +39,19 @@ function rrule(::typeof(hcat), A::AbstractArray, Bs::AbstractArray...)
     return hcat(A, Bs...), hcat_pullback
 end
 
+function rrule(::typeof(reduce), ::typeof(hcat), As::AbstractVector{<:AbstractVecOrMat})
+    function reduce_hcat_pullback(Ȳ)
+        sizes = size.(As, 2)
+        cumsizes = cumsum(sizes)
+        ∂As = map(cumsizes, sizes) do post, diff
+            pre = post - diff + 1
+            Ȳ[:, pre:post]
+        end
+        return (NO_FIELDS, NO_FIELDS, ∂As)
```

Comment:
I think `DoesNotExist()` should be used here, as `NO_FIELDS` is reserved for the calling function. This is what we do for e.g. `sum(abs, x)`, but I could be wrong about this:

```suggestion
        return (NO_FIELDS, DoesNotExist(), ∂As)
```

Reply:
This is the one style change where I'm not sure I agree.

```julia
help?> NO_FIELDS
search: NO_FIELDS

  NO_FIELDS

  Constant for the reverse-mode derivative with respect to a structure that
  has no fields. The most notable use for this is for the reverse-mode
  derivative with respect to the function itself, when that function is not a
  closure.

help?> DoesNotExist
search: DoesNotExist

  DoesNotExist() <: AbstractZero

  This differential indicates that the derivative does not exist. It is the
  differential for a Primal type that is not differentiable. Such an Integer,
  or Boolean (when not being used as a represention of a value that normally
  would be a floating point.)
```

A function (not a closure) is either "struct with no fields" or "not differentiable primal type": how can it be one as first argument to `rrule`, and the other as second argument?